1.7.0 Flag 'mark_files_as_generated' and refactoring

StarProxima opened the pull request:
It might be worth adding a pretty comment for the generated files to indicate that they are generated, indicate the source of generation and disable linter for them.

Example:

```dart
//  ____ _ _ _ ____ ____ ____ ____ ____     ___  ____ ____ ____ ____ ____
//  [__  | | | |__| | __ | __ |___ |__/     |__] |__| |__/ [__  |___ |__/
//  ___] |_|_| |  | |__] |__] |___ |  \ ___ |    |  | |  \ ___] |___ |  \                            
//
// GENERATED CODE - DO NOT MODIFY BY HAND
// ignore_for_file: type=lint
    
import 'package:dio/dio.dart';

import 'client/client_client.dart';

abstract class IRestClient {
  ClientClient get client;
}

/// Petstore `v1.0`
/// 
/// A sample API that uses a petstore as an example to demonstrate features in the OpenAPI 3.0 specification
class RestClient implements IRestClient {
  RestClient({
    required Dio dio,
    required String baseUrl,
  })  : _dio = dio,
        _baseUrl = baseUrl;

  final Dio _dio;
  final String _baseUrl;

  ClientClient? _client;

  @override
  ClientClient get client => _client ??= ClientClient(_dio, baseUrl: _baseUrl);
}
```

Carapacik commented:
I'm not sure that this is necessary, because these files are not created by build_runner

StarProxima commented:
> I'm not sure that this is necessary, because these files are not created by build_runner

Yes, but does it make a difference to the end user? These generated files are not meant to be changed, as they can be re-generated the next time swagger_parser is run and the changes can be lost. 

I think that in a team, a developer who has not personally configured the generation through swagger_parser might get the false impression that these files are handwritten.

Wouldn't it be better to specify that these files are generated so that developers, especially not very experienced ones, don't have problems with losing changes? Also, such a comment disables linter, which fixes problems with sorting imports, for example.

If it is still necessary to make changes in the code, this comment will make it clear that it is better to remove these files from this directory. 

It seems to me that such a comment provides more clarity.

Sancene commented:
These files are not meant to be generated by all project members really but rather are just a quick setup for existing swagger schema so files should be committed and be perceived as handwritten. I do not believe that this change is necessary especially the linter ignore part. We can make a flag that is defaulted to true for generated code comment but i dont see a reason why.

StarProxima commented:
> These files are not meant to be generated by all project members really but rather are just a quick setup for existing swagger schema so files should be committed and be perceived as handwritten. I do not believe that this change is necessary especially the linter ignore part. We can make a flag that is defaulted to true for generated code comment but i dont see a reason why.

There are cases where the backend is under development, often new methods are added and existing methods are modified, names and models are changed.

When this happens, it is crucial not to make manual changes to code that will change quite often over time, so a comment like this would be extremely helpful.

Ignoring the linker in such a case also has advantages, because the analyzer will not take into account the problems that may arise with conflicting linker rules (always_use_package_imports, directives_ordering, etc) in the code that you are not responsible for, because it is generated.

That's why I think it's worth adding an optional flag to the configuration as you suggested.

If you confirm this solution, then I will try to finish the pull request today.

StarProxima commented:
@Carapacik @Sancene could you please take a look?

I've added an optional flag to generate a comment in the files.

Also small improvements on the code in FillController and Generator so that the default config parameters are only set in one place, fields are now immutable.

Also cleaned up the code in YamlConfig, reduced unnecessary checks, fields are now immutable.

Carapacik commented on swagger_parser/CHANGELOG.md:
```diff
@@ -1,3 +1,14 @@
+## 1.6.4
+- Add new config parameter `mark_files_as_generated`
```
Add version 1.7.0 for this and include all other fixes

StarProxima replied:
Edited in https://github.com/Carapacik/swagger_parser/pull/90/commits/b4ff6e95e2ff5a67979bd72fd4951282b67df4ad

Carapacik commented on swagger_parser/lib/src/config/yaml_config.dart:
```diff
@@ -14,7 +15,24 @@ import 'config_exception.dart';
-  YamlConfig(List<String> arguments) {
+  /// Applies parameters directly from constructor
+  const YamlConfig({
+    required this.schemaFilePath,
```
Please remove required for all fields in constructor

StarProxima replied:
required in the constructor makes it mandatory to pass all parameters to the factory constructor, which can prevent errors when adding new configuration parameters when fromYamlFile does not pass all parameters to YamlConfig. In general, this constructor can be made private, since it is only used for internal purposes.

Carapacik commented on swagger_parser/lib/src/utils/utils.dart:
```diff
@@ -48,6 +48,25 @@ String ioImport(UniversalComponentClass dataClass) => dataClass.parameters
     ? "import 'dart:io';\n\n"
     : '';
 
+const dartGeneratedFileComment = r'''
```
Separate this text to generated file comment and ignore_for_file comment

StarProxima replied:
How do you want me to separate them and how do you want me to use them in templates? If using them separately we don't need a config parameter for ignore_for_file?

Carapacik replied:
Instead of dartGeneratedFileComment and kotlinGeneratedFileComment, use generated File Comment and ignore file comment (which is used only in dart)

Carapacik commented on swagger_parser/lib/src/utils/utils.dart:
```diff
@@ -48,6 +48,25 @@ String ioImport(UniversalComponentClass dataClass) => dataClass.parameters
     : '';
 
+const dartGeneratedFileComment = r'''
+//  ____ _ _ _ ____ ____ ____ ____ ____     ___  ____ ____ ____ ____ ____
```
Also change generated text to 
```dart
// coverage:ignore-file
// GENERATED CODE - DO NOT MODIFY BY HAND
// ignore_for_file: type=lint
```

StarProxima replied:
Edited in https://github.com/Carapacik/swagger_parser/pull/90/commits/8829f6263cf6f2bffd7526dead686398b33e5532

Carapacik commented on swagger_parser/pubspec.yaml:
```diff
@@ -1,6 +1,6 @@
 name: swagger_parser
 description: Package that generates REST clients and data classes from OpenApi definition file
-version: 1.6.1
+version: 1.6.4
```
1.7.0

StarProxima replied:
Edited in https://github.com/Carapacik/swagger_parser/pull/90/commits/b4ff6e95e2ff5a67979bd72fd4951282b67df4ad

Carapacik commented on swagger_parser/lib/src/generator/fill_controller.dart:
```diff
@@ -9,25 +9,45 @@ import 'models/universal_rest_client.dart';
+        _markFilesAsGenerated = markFilesAsGenerated;
+
+  /// Constructor for tests that accepts configuration parameters with default values to create files
+  const FillController.forTests({
```
Please remove this constructor, there is no need to add a constructor with all required fields

StarProxima replied:
In this case, a constructor with mandatory fields is needed both to ensure that all parameters are passed, and to set all default values of configuration parameters in the Generator constructor so that only one axis of changes remains when default values are changed.

A separate constructor for tests is needed in order not to pass all parameters each time, but to set them as defaults only for testing purposes. For example, it makes no sense to set the `mark_files_as_generated` flag as true in tests.

Perhaps it should be renamed, but the current name seems to me to explain the essence of its use.

Carapacik commented on swagger_parser/lib/src/generator/generator.dart:
```diff
@@ -16,126 +16,106 @@ import 'models/universal_rest_client.dart';
+  Generator({
+    required String schemaContent,
+    required String outputDirectory,
+    required ProgrammingLanguage? language,
```
Why not to do 
```ProgrammingLanguage language = ProgrammingLanguage.dart```
with all fields, instead of doing it after the constructor?

StarProxima replied:
All arguments are marked as required for the same reasons I described above, 
This constructor is used by the Generator.fromYamlConfig factory constructor, which because of the mandatory nullable arguments can pass parameters directly from yamlConfig without checking each parameter for null, and each null parameter passed will be replaced with the default parameter.

Also, since swagger_parser_pages uses this constructor, it's a good idea to make sure we didn't miss any of the parameters when passing them.

Also, swagger_parser_pages can pass null to some fields, such as clientPostfix, so we need to set default values in the initializers to avoid setting them on the swagger_parser_pages side.

StarProxima commented:
@Carapacik I fixed some of the comments and left a comment on the rest. If you still have any doubts, please let me know.

Sancene commented on swagger_parser/lib/src/generator/templates/kotlin_typedef_template.dart:
```diff
@@ -7,12 +7,15 @@ import '../models/universal_data_class.dart';
-String kotlinTypeDefTemplate(UniversalComponentClass dataClass) {
+String kotlinTypeDefTemplate(
+  UniversalComponentClass dataClass, {
+  required bool markFileAsGenerated,
```
why have it as required instead of defaulting it to false?
